### PR TITLE
CP2K+SMEAGOL interface for electron transport with non-equilibrium Green’s functions

### PR DESCRIPTION
**Pull request by chrisahart**
A new interface between CP2K and the non-equilibrium Green’s functions (NEGF) code SMEAGOL, as recently published in the Journal of Chemical Theory and Computation.

The SMEAGOL code can be downloaded by contacting Dr. Clotilde Cucinotta at cucinotta@imperial.ac.uk or Prof. Stefano Sanvito at sanvitos@tcd.ie.

Ahart, C. S., Chulkov, S. K. & Cucinotta, C. S. Enabling Ab-Initio Molecular Dynamics under Bias: The CP2K+SMEAGOL Interface for Integrating Density Functional Theory and Non-Equilibrium Green Functions. J. Chem. Theory Comput. 2024.

### CODE REVIEW

**Comment by oschuett** [comment]
This PR is tricky.

First of all, I do recognize that you put **a lot** of work into this.

However, I'm very concerned about the fact that SMEAGOL is only available upon personal request. This is not in the spirit of open source and the GPL in particular. 

Is there perhaps a way to make SMEAGOL accessible to the general public?

**Comment by chrisahart** [comment]
Hi Ole, thank you for your comment. I agree that it is problematic that the SMEAGOL code is only available upon personal request. 

We have developed an interface between CP2K and SMEAGOL, and do not currently have permission to distribute the SMEAGOL code itself. 

The developers of the SMEAGOL code have agreed to move forward with an open source license in the near future. Once this has been finalized we will submit a new pull request with a publicly available download link for SMEAGOL.

**Comment by schulkov** [comment]
Hi Ole, historically SMEAGOL has been distributing according to Siesta Academic License terms. This is because its source code is tightly coupled with an ancient Siesta 1.3f source code, which is still subject to the proprietary Siesta Academic License (https://www.siesta-project.org/SIESTA_MATERIAL/OlderCode/).

As part of developing this CP2K+SMEAGOL interface, SMEAGOL source code was decoupled from Siesta 1.3f and converted into a library. Obtaining consents from all major contributors to the SMEAGOL project is the only thing that prevents releasing this SMEAGOL library under the terms of an open source license. No one has strictly objected against it so far.

**Comment by oschuett** [comment]
@schulkov  thanks a lot for the additional context!
Given these special circumstances, I think we can merge this PR without creating a precedent.

PS: Kudos for taking on that license change.

File: src/input_cp2k_smeagol.F [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+      !        BulkLead, NIMAGES_IJ, EnergyGridType, TypeOfIntegral
+      CALL section_create(section, __LOCATION__, name="SMEAGOL", &
+                          description="Parameters for NEGF/SMEAGOL simulation.", &
+                          n_keywords=${len(reademtr_local_lnames)+len(reademtr_negfmod_lnames)+len(reademtr_negfcoop_lnames)+len(reademtr_local_inames)+len(reademtr_negfmod_inames)+len(reademtr_negfcoop_inames)+len(reademtr_local_rnames)+len(reademtr_negfmod_rnames)+len(reademtr_local_pnames)+len(reademtr_negfmod_rydberg_pnames)+len(reademtr_negfmod_pnames)+len(readoptsnegf_negfmod_lnames)+len(readoptsnegf_negfmod_inames)+len(readoptsnegf_negfmod_rnames)+len(readoptsnegf_negfmod_explicit_pnames)+len(readoptsnegf_bfield_rydberg_pnames)+len(emtoptions_negfmod_lnames)+len(emtoptions_negfmod_inames)+len(emtoptions_local_inames)+len(emtoptions_local_explicit_inames)+len(emtoptions_negfmod_explicit_inames)+len(emtoptions_sigma_inames)+len(emtoptions_negfmod_rnames)+len(emtoptions_negfmod_explicit_rnames)+len(emtoptions_negfmod_rydberg_pnames)+len(smeagol_interface_local_pnames)+11}$, &
```
Comment (oschuett) [inline src/input_cp2k_smeagol.F]: The `n_keywords` argument is actually optional (IMHO it's a premature optimization). Feel free to omit it.

**Comment by cucinotta** [comment]
I have just been informed that the Smeagol developers are planning to make the repository containing the Smeagol library publicly accessible today

**Comment by chrisahart** [comment]
SMEAGOL is now available under GPL license at https://github.com/StefanoSanvitoGroup/libsmeagol

**Comment by oschuett** [comment]
It's awesome that SMEAGOL is now public!

In principle this PR is ready to be merged. Obviously, without actually installing SMEAGOL in the toolchain, we can't test it. So it would be great if you could follow up with a later PR once SMEAGOL has made its first official release.

My only remaining remark would be that most of the new keywords [lack documentation](https://ci.cp2k.org/artifacts/run-cp2k-manual-d8f3927b/manual/CP2K_INPUT/FORCE_EVAL/DFT/SMEAGOL.html).

**Comment by chrisahart** [comment]
As Sergey explained we converted SMEAGOL into a library that can be linked to CP2K, and therefore have chosen to provide support for all SMEAGOL keywords. The SMEAGOL user guide documents some of these keywords https://github.com/StefanoSanvitoGroup/libsmeagol/blob/main/Docs_Smeagol/Smeagol-1.2.pdf

We have documented any keywords that are specific to CP2K+SMEAGOL such as RUN_TYPE, NIMAGES_IJ and PROJECT_NAME.

Hopefully in the future any obsolete keywords will be removed and the remaining keywords properly documented, however I feel this is outside the scope of the current work.

**Comment by oschuett** [comment]
> Hopefully in the future any obsolete keywords will be removed and the remaining keywords properly documented, ...

Yes, since CP2K keywords are case insensitive, it would be good to switch from Camel Case to Snake Case. Furthermore, one could introduce subsections for common prefixes such as  `AM`, `BS` ,`EM`, `SIGMA`, etc.

> ... however I feel this is outside the scope of the current work.

I agree, one has to draw the line somewhere. Let's merge :-)

**Pull request merged**